File: kata/bowling_game/rust/2024_01_13/bowling_game/tests/bowling_game_test.rs
use bowling_game::Game;

#[cfg(test)]
mod tests {
    use super::*;

    fn roll_many(game: &mut Game, n: u16, pins: u16) {
        for _ in 0..n {
            game.roll(pins)
        }
    }

    fn roll_standard_spare(game: &mut Game) {
        game.roll(5);
        game.roll(5);
    }

    fn roll_different_spare(game: &mut Game) {
        game.roll(3);
        game.roll(7);
    }

    fn roll_strike(game: &mut Game) {
        game.roll(10);
    }

    #[test]
    fn test_gutter_game() {
        let mut game = Game::new();
        roll_many(&mut game, 20, 0);
        assert_eq!(game.score(), 0);
    }

    #[test]
    fn test_all_ones() {
        let mut game = Game::new();
        roll_many(&mut game, 20, 1);
        assert_eq!(game.score(), 20);
    }

    #[test]
    fn test_standard_spare() {
        let mut game = Game::new();
        roll_standard_spare(&mut game);
        game.roll(3);
        roll_many(&mut game, 17, 0);
        assert_eq!(game.score(), 16);
    }

    #[test]
    fn test_different_spare() {
        let mut game = Game::new();
        roll_different_spare(&mut game);
        game.roll(3);
        roll_many(&mut game, 17, 0);
        assert_eq!(game.score(), 16);
    }

    #[test]
    fn test_one_strike() {
        let mut game = Game::new();
        roll_strike(&mut game);
        game.roll(3);
        game.roll(4);
        roll_many(&mut game, 16, 0);
        assert_eq!(game.score(), 24);
    }
}

File: kata/bowling_game/go/2024_02_03/bowling_game_test.go
package bowling

import "testing"

var game *Game

func setUp() {
	game = NewGame()
}

func rollMany(n int, pins int) {
	for i := 0; i < n; i++ {
		game.Roll(pins)
	}
}

func rollStandardSpare() {
	game.Roll(5)
	game.Roll(5)
}

func rollDifferentSpare() {
	game.Roll(1)
	game.Roll(9)
}

func rollStrike() {
	game.Roll(10)
}

func TestGutterGame(t *testing.T) {
	setUp()

	rollMany(20, 0)

	if game.Score() != 0 {
		t.Errorf("Expected score of 0, but got %d", game.Score())
	}
}

func TestAllOnes(t *testing.T) {
	setUp()

	rollMany(20, 1)

	if game.Score() != 20 {
		t.Errorf("Expected score of 20, but got %d", game.Score())
	}
}

func TestStandardSpare(t *testing.T) {
	setUp()

	rollStandardSpare()
	game.Roll(3)
	rollMany(17, 0)

	if game.Score() != 16 {
		t.Errorf("Expected score of 16, but got %d", game.Score())
	}
}

func TestDifferentSpare(t *testing.T) {
	setUp()

	rollDifferentSpare()
	game.Roll(3)
	rollMany(17, 0)

	if game.Score() != 16 {
		t.Errorf("Expected score of 16, but got %d", game.Score())
	}
}

func TestOneStrike(t *testing.T) {
	setUp()

	rollStrike()
	game.Roll(3)
	game.Roll(4)
	rollMany(16, 0)

	if game.Score() != 24 {
		t.Errorf("Expected score of 24, but got %d", game.Score())
	}
}

func TestPerfectGame(t *testing.T) {
	setUp()

	rollMany(12, 10)

	if game.Score() != 300 {
		t.Errorf("Expected score of 300, but got %d", game.Score())
	}
}
